File: src/binding/csharp/tests/ZVec.Tests/ZVecTests.cs
// Copyright 2025-present the zvec project
using System;
using System.IO;
using System.Collections.Generic;
using Xunit;

namespace ZVec.Tests
{
    public class ZVecTests : IDisposable
    {
        private readonly List<string> _dirs = new();

        private string TestDir(string name)
        {
            string dir = $"/tmp/zvec_csharp_test_{name}";
            if (Directory.Exists(dir)) Directory.Delete(dir, true);
            _dirs.Add(dir);
            return dir;
        }

        public void Dispose()
        {
            foreach (var dir in _dirs)
                if (Directory.Exists(dir)) Directory.Delete(dir, true);
        }

        [Fact]
        public void TestSchemaCreate()
        {
            using var schema = new Schema("test_schema");
            schema.AddField("vec", DataType.VectorFP32, 4);
            schema.AddField("name", DataType.String);
            schema.AddField("age", DataType.Int32);
        }

        [Fact]
        public void TestDocLifecycle()
        {
            using var doc = new Doc();
            doc.SetPK("pk_001");
            Assert.Equal("pk_001", doc.PK());
            doc.SetString("name", "Alice");
            doc.SetInt32("age", 30);
            doc.SetVector("vec", new float[] { 0.1f, 0.2f, 0.3f, 0.4f });
        }

        [Fact]
        public void TestCollectionCRUD()
        {
            var dir = TestDir("crud");
            using var schema = new Schema("crud_col");
            schema.AddField("vec", DataType.VectorFP32, 4);
            schema.AddField("name", DataType.String);

            using var col = new Collection(dir, schema);
            using var doc = new Doc();
            doc.SetPK("u1");
            doc.SetVector("vec", new float[] { 1, 2, 3, 4 });
            doc.SetString("name", "Bob");
            col.Upsert(new[] { doc });
            col.Flush();

            Assert.Equal(1UL, col.Stats());

            var results = col.Fetch(new[] { "u1" });
            Assert.Single(results);
            Assert.Equal("u1", results[0].PK());
            foreach (var d in results) d.Dispose();
        }

        [Fact]
        public void TestInvalidOpen()
        {
            Assert.Throws<ZVecException>(() => new Collection("/tmp/zvec_csharp_nonexistent_abc123"));
        }

        [Fact]
        public void TestBatchUpsert()
        {
            var dir = TestDir("batch");
            using var schema = new Schema("batch_col");
            schema.AddField("vec", DataType.VectorFP32, 4);

            using var col = new Collection(dir, schema);
            var docs = new Doc[100];
            for (int i = 0; i < 100; i++)
            {
                docs[i] = new Doc();
                docs[i].SetPK($"batch_{i}");
                docs[i].SetVector("vec", new float[] { i, i, i, i });
            }
            col.Upsert(docs);
            col.Flush();
            Assert.Equal(100UL, col.Stats());
            foreach (var d in docs) d.Dispose();
        }

        [Fact]
        public void TestFetchNonexistentPK()
        {
            var dir = TestDir("fetchnone");
            using var schema = new Schema("fetchnone_col");
            schema.AddField("vec", DataType.VectorFP32, 4);

            using var col = new Collection(dir, schema);
            var results = col.Fetch(new[] { "does_not_exist" });
            Assert.Empty(results);
        }
    }
}

File: src/binding/swift/Tests/ZVecTests/ZVecTests.swift
// Copyright 2025-present the zvec project
import XCTest
@testable import ZVec

final class ZVecTests: XCTestCase {

    func testDir(_ name: String) -> String {
        let dir = "/tmp/zvec_swift_test_\(name)"
        try? FileManager.default.removeItem(atPath: dir)
        return dir
    }

    func testSchemaCreate() throws {
        let schema = Schema(name: "test_schema")
        try schema.addField(name: "vec", type: ZVEC_TYPE_VECTOR_FP32, dimension: 4)
        try schema.addField(name: "name", type: ZVEC_TYPE_STRING)
        try schema.addField(name: "age", type: ZVEC_TYPE_INT32)
    }

    func testDocLifecycle() throws {
        let doc = Doc()
        doc.setPK("pk_001")
        XCTAssertEqual(doc.pk(), "pk_001")

        try doc.setString(field: "name", value: "Alice")
        try doc.setInt32(field: "age", value: 30)
        try doc.setVector(field: "vec", vector: [0.1, 0.2, 0.3, 0.4])
    }

    func testCollectionCRUD() throws {
        let dir = testDir("crud")
        let schema = Schema(name: "crud_col")
        try schema.addField(name: "vec", type: ZVEC_TYPE_VECTOR_FP32, dimension: 4)
        try schema.addField(name: "name", type: ZVEC_TYPE_STRING)

        let col = try Collection(path: dir, schema: schema)

        // Upsert
        let doc = Doc()
        doc.setPK("u1")
        try doc.setVector(field: "vec", vector: [1.0, 2.0, 3.0, 4.0])
        try doc.setString(field: "name", value: "Bob")
        try col.upsert(docs: [doc])
        try col.flush()

        // Stats
        let count = try col.stats()
        XCTAssertEqual(count, 1)

        // Fetch
        let results = try col.fetch(pks: ["u1"])
        XCTAssertEqual(results.count, 1)
        XCTAssertEqual(results[0].pk(), "u1")
    }

    func testInvalidOpen() {
        XCTAssertThrowsError(try Collection(path: "/tmp/zvec_swift_nonexistent_abc123"))
    }

    func testBatchUpsert() throws {
        let dir = testDir("batch")
        let schema = Schema(name: "batch_col")
        try schema.addField(name: "vec", type: ZVEC_TYPE_VECTOR_FP32, dimension: 4)

        let col = try Collection(path: dir, schema: schema)

        var docs: [Doc] = []
        for i in 0..<100 {
            let d = Doc()
            d.setPK("batch_\(i)")
            try d.setVector(field: "vec", vector: [Float(i), Float(i), Float(i), Float(i)])
            docs.append(d)
        }
        try col.upsert(docs: docs)
        try col.flush()

        let count = try col.stats()
        XCTAssertEqual(count, 100)
    }

    func testFetchNonexistentPK() throws {
        let dir = testDir("fetchnone")
        let schema = Schema(name: "fetchnone_col")
        try schema.addField(name: "vec", type: ZVEC_TYPE_VECTOR_FP32, dimension: 4)

        let col = try Collection(path: dir, schema: schema)

        let results = try col.fetch(pks: ["does_not_exist"])
        XCTAssertEqual(results.count, 0)
    }
}
